sqlmesh/utils/dag.py

```diff
@@ -80,23 +83,25 @@ def graph(self) -> t.Dict[T, t.Set[T]]:
             graph[node] = deps.copy()
         return graph
 
+    @property
     def sorted(self) -> t.List[T]:
         """Returns a list of nodes sorted in topological order."""
-        result: t.List[T] = []
+        if self._sorted is None:
+            self._sorted = []
 
-        unprocessed_nodes = self.graph
-        while unprocessed_nodes:
-            next_nodes = {node for node, deps in unprocessed_nodes.items() if not deps}
+            unprocessed_nodes = self.graph
+            while unprocessed_nodes:
+                next_nodes = {node for node, deps in unprocessed_nodes.items() if not deps}
```

Comment:
We should raise an exception if next_nodes is empty. It means we have a circular dependency between the remaining unprocessed nodes.